### PR TITLE
ctransformers improvements

### PR DESCRIPTION
**Pull request by cal066**
1. Context length setting.
2. Better "None" type guessing since infer loader does not set model_type.
3. Stop using _stream internal function calls.
## Checklist:

- [x] I have read the [Contributing guidelines](https://github.com/oobabooga/text-generation-webui/wiki/Contributing-guidelines).

### CODE REVIEW

**Comment by cal066** [comment]
Hello @netrunnereve Looks like we have been working in parallel, I saw your #3556. Do you mind rebasing the changes on top of mine?

@oobabooga I found a problem after a1a9ec895d96b27f2c03cf5df4bce679f3abaf91. I'm somehow getting:
```
text-generation-webui  | Traceback (most recent call last):                                                                                                                                                                                     
text-generation-webui  |   File "/app/modules/callbacks.py", line 55, in gentask                                                                                                                                                                
text-generation-webui  |     ret = self.mfunc(callback=_callback, *args, **self.kwargs)                                                                                                                                                         
text-generation-webui  |   File "/app/modules/ctransformers_model.py", line 94, in generate                                                                                                                                                     
text-generation-webui  |     for token in generator:                                                                                                                                                                                            
text-generation-webui  |   File "/venv/lib/python3.10/site-packages/ctransformers/llm.py", line 467, in _stream                                                                                                                                 
text-generation-webui  |     for token in self.generate(                                                                                                                                                                                        
text-generation-webui  |   File "/venv/lib/python3.10/site-packages/ctransformers/llm.py", line 426, in generate                                                                                                                                
text-generation-webui  |     token = self.sample(                                                                                                                                                                                               
text-generation-webui  |   File "/venv/lib/python3.10/site-packages/ctransformers/llm.py", line 377, in sample                                                                                                                                  
text-generation-webui  |     return self.ctransformers_llm_sample(                                                                                                                                                                              
text-generation-webui  | ctypes.ArgumentError: argument 7: TypeError: wrong type                                                                                                                                                                
text-generation-webui  | Output generated in 6.04 seconds (0.00 tokens/s, 0 tokens, context 19, seed 1883392458)
```

Somehow state['seed'] ended up as a float??
```
logger.info(f'type is {type(state["seed"])}, value: {state["seed"]}')
INFO:type is <class 'float'>, value: -1.0
```
I used `seed=int(state['seed'])` locally as a workaround.

**Comment by cal066** [comment]
I have decided to follow `set_manual_seed` as an example to force an integer. Is there a way to actually get the value from `set_manual_seed` in the generate call?

File: modules/ctransformers_model.py [hunk 1]
```diff
@@ -9,29 +12,57 @@ class CtransformersModel:
...
-        return shared.args.model_type == "Auto" or shared.args.model_type == "None"
+        return shared.args.model_type is None or shared.args.model_type == "Auto" or shared.args.model_type == "None"
+
+    # llama is more of a fallback option
+    match_table = OrderedDict({
+        'falcon': re.compile('.*falcon.*ggml.*\\.bin'),
+        'mpt': re.compile('.*mpt.*ggml.*\\.bin'),
+        'starcoder': re.compile('.*star(chat|coder).*ggml.*\\.bin'),
+        'llama': re.compile('.*ggml.*\\.bin')
+    })
```
Comment (oobabooga) [inline modules/ctransformers_model.py:45]: The webui has its own mechanism for inferring and setting parameters based on model names. The patterns are at `models/config.yaml`, and there you can set `model_type` based on regexes. So this should be removed and moved there.

File: modules/ctransformers_model.py [hunk 2]
```diff
@@ -9,29 +12,57 @@ class CtransformersModel:
     def __init__(self):
         pass
 
-    @classmethod
```
Comment (oobabooga) [inline modules/ctransformers_model.py:12]: Why remove the classmethod? Is there an advantage?

File: modules/ctransformers_model.py [hunk 3]
```diff
@@ -9,29 +12,57 @@ class CtransformersModel:
...
+            return None
+
+        # Try guessing
+        file_list = list(model_name_dir.glob('*.bin'))
+        if len(file_list) > 0 and file_list[0] is not None:
+            for type_match in self.match_table.keys():
+                pattern = self.match_table.get(type_match)
+                if(pattern.match(str(file_list[0].name))):
+                    logger.info(f'Guessing {file_list[0].name} is {type_match}')
+                    return type_match
```
Comment (oobabooga) [inline modules/ctransformers_model.py:62]: This part should be removed, see the comment above on `models/config.yaml`.

**Comment by oobabooga** [comment]
> I have decided to follow `set_manual_seed` as an example to force an integer. Is there a way to actually get the value from `set_manual_seed` in the generate call?

That's geared towards pytorch. I think it's fine to use the UI seed for ctransformers as it is for the time being.

File: modules/ctransformers_model.py (hunk 2 above)
Comment (cal066) [reply modules/ctransformers_model.py:12]: It's for consistency with the other methods that use self. The classmethod wasn't intended.

File: modules/ctransformers_model.py (hunk 3 above)
Comment (cal066) [reply modules/ctransformers_model.py]: Done.

File: modules/ctransformers_model.py (hunk 1 above)
Comment (cal066) [reply modules/ctransformers_model.py]: I have dropped this part in favor of the more conservative `models/config.yaml` changes.

**Comment by oobabooga** [comment]
I merged the rebased version at https://github.com/oobabooga/text-generation-webui/pull/3556. Thanks @cal066!

**Pull request closed without merging**